File: android/src/main/java/com/reveila/android/service/VaultScannerWorker.java
package com.reveila.android.service;

import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.core.app.NotificationCompat;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import com.reveila.android.ModelSettings;
import com.reveila.android.ScanResult;
import com.reveila.android.VaultScanner;

/**
 * Java implementation of the Background scanner.
 * Decoupled from Kotlin Coroutines and specific UI Activities.
 */
public class VaultScannerWorker extends Worker {

    private static final String TAG = "VaultScannerWorker";
    private static final String CHANNEL_ID = "REVEILA_MEM_CHANNEL";

    public VaultScannerWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
        super(context, workerParams);
    }

    @NonNull
    @Override
    public Result doWork() {
        // Assume ModelSettings is already converted to Java or is a simple POJO
        ModelSettings settings = new ModelSettings(getApplicationContext());
        String vaultUriString = settings.getVaultUri();
        
        if (vaultUriString == null || vaultUriString.isEmpty()) {
            return Result.success();
        }

        String keywords = settings.getFocusKeywords();

        Log.i(TAG, "Starting periodic background scan for: " + vaultUriString);

        try {
            VaultScanner scanner = new VaultScanner(getApplicationContext());
            Uri uri = Uri.parse(vaultUriString);
            
            // Note: Repository logic should be added here once VaultScanner 
            // is initialized with your Java Room stack.
            // For now, we pass null for the listener/repo to keep it headless.
            ScanResult result = scanner.performScan(uri, null, keywords, null);
            
            if (result.getNewFilesCount() > 0) {
                notifyUser(result.getNewFilesCount(), result.getEntitiesDiscoveredCount());
            }
            
            return Result.success();
        } catch (Exception e) {
            Log.e(TAG, "Background scan failed", e);
            return Result.retry();
        }
    }

    private void notifyUser(int newFiles, int totalEntities) {
        NotificationManager notificationManager = (NotificationManager) 
                getApplicationContext().getSystemService(Context.NOTIFICATION_SERVICE);
        
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel = new NotificationChannel(
                CHANNEL_ID,
                "Sovereign Memory Updates",
                NotificationManager.IMPORTANCE_LOW
            );
            channel.setDescription("Notifies when the local AI knowledge graph is updated.");
            notificationManager.createNotificationChannel(channel);
        }

        // REDIRECT: Open the main launcher activity (Expo Shell) instead of SetupActivity
        Intent intent = getApplicationContext().getPackageManager()
                .getLaunchIntentForPackage(getApplicationContext().getPackageName());
        
        if (intent != null) {
            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
        }

        PendingIntent pendingIntent = PendingIntent.getActivity(
            getApplicationContext(), 
            0, 
            intent, 
            PendingIntent.FLAG_IMMUTABLE | PendingIntent.FLAG_UPDATE_CURRENT
        );

        NotificationCompat.Builder builder = new NotificationCompat.Builder(getApplicationContext(), CHANNEL_ID)
            .setSmallIcon(android.R.drawable.ic_popup_sync) 
            .setContentTitle("Sovereign Memory Updated")
            .setContentText("Internalized " + newFiles + " new documents and " + totalEntities + " key entities.")
            .setPriority(NotificationCompat.PRIORITY_LOW)
            .setContentIntent(pendingIntent)
            .setAutoCancel(true);

        notificationManager.notify(1001, builder.build());
    }
}

File: android/src/main/kotlin/com/reveila/android/service/VaultScannerWorker.kt
package com.reveila.android.service

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.util.Log
import androidx.core.app.NotificationCompat
import androidx.work.CoroutineWorker
import androidx.work.WorkerParameters
import com.reveila.android.ModelSettings
import com.reveila.android.SovereignSetupActivity
import com.reveila.android.VaultScanner

/**
 * Background worker that periodically performs a Delta Scan on the Knowledge Vault.
 */
class VaultScannerWorker(
    context: Context,
    params: WorkerParameters
) : CoroutineWorker(context, params) {

    override suspend fun doWork(): Result {
        val settings = ModelSettings(applicationContext)
        val vaultUriString = settings.vaultUri ?: return Result.success()
        val keywords = settings.focusKeywords

        Log.i("VaultScannerWorker", "Starting periodic background scan for: $vaultUriString")

        return try {
            val scanner = VaultScanner(applicationContext)
            val uri = Uri.parse(vaultUriString)
            
            // Background scan doesn't have a UI ViewModel to update
            val result = scanner.performScan(uri, null, keywords)
            
            if (result.newFilesCount > 0) {
                notifyUser(result.newFilesCount, result.entitiesDiscoveredCount)
            }
            
            Result.success()
        } catch (e: Exception) {
            Log.e("VaultScannerWorker", "Background scan failed", e)
            Result.retry()
        }
    }

    private fun notifyUser(newFiles: Int, totalEntities: Int) {
        val notificationManager = applicationContext.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        
        // Create Notification Channel for API 26+
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                "REVEILA_MEM_CHANNEL",
                "Sovereign Memory Updates",
                NotificationManager.IMPORTANCE_LOW
            ).apply {
                description = "Notifies when the local AI knowledge graph is updated."
            }
            notificationManager.createNotificationChannel(channel)
        }

        val intent = Intent(applicationContext, SovereignSetupActivity::class.java).apply {
            flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
        }
        val pendingIntent = PendingIntent.getActivity(
            applicationContext, 
            0, 
            intent, 
            PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT
        )

        val notification = NotificationCompat.Builder(applicationContext, "REVEILA_MEM_CHANNEL")
            .setSmallIcon(android.R.drawable.ic_popup_sync) // Fallback icon if ic_reveila_watchdog is missing
            .setContentTitle("Sovereign Memory Updated")
            .setContentText("Internalized $newFiles new documents and $totalEntities key entities.")
            .setPriority(NotificationCompat.PRIORITY_LOW) // Stay non-intrusive
            .setContentIntent(pendingIntent)
            .setAutoCancel(true)
            .build()

        notificationManager.notify(1001, notification)
    }
}
